### PR TITLE
Draft of Multimic using policy language.

### PR DESCRIPTION
**Pull request by markww**
This is a draft of the multimic example that tries to express the policy
enforcement preferences in the authorization logic language. This does
not currently build, as I'm not 100% sure how to express what I want in
the authorization logic language, so putting up a draft for some help.

### CODE REVIEW

**Comment by aferr** [comment]
> I'm also trying to do some things that I'm not sure if the language supports. In essence, I am having the 
> `DevicePolicyAuthority` only grant permissions to users that have granted it some permissions first. This 
> "permissions handshake" seems like a good pattern in a world of distributed trust, but I may be running over  > some of the guardrails of the language.

OK I can try to help think about how to do that. A policy handshake seems like a good thing to try to cover. At the moment, the allocation of resources to stakeholders in the code looks a bit different from what I would expect. I could be misunderstanding how your principals map to real-world stakeholders though.

Here's my assumption, correct me if this is wrong: I think "DevicePolicyAuthority" is either a policy set by the device/service manufacturer, or similarly it's an open-source policy standard that this service/device provider will apply or delegate to.

Here's how I think the allocation of resources and handshake would go:
    - The users in the room own their speech and derivations of it (like transcripts)
    - The service provider owns its services and its proprietary data like ML models
    - The handshake happens because:
          - The service provider is only willing to let humans use its services models if it get to do <something> with <some derivative of> the user data.
          - The users are only interested in these services if they'll respect their privacy needs/preferences.

In the policy code it looks a bit more like the "DevicePolicyAuthority" owns the data to begin with, so if this is true it seems like it can just do what it wants to begin with. I might be confused about this though.

File: src/analysis/souffle/examples/multimic/multimic.authlogic [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+  // This policy authority is promising that it will always notify when it has
+  // to override a user preference.
+  notifiesOnPreferenceOverride(tagX) :- isCollaborationTag(tagX).
+
+  // The DevicePolicyAuthority will only grant permissions to users to
+  // influence collaboration tags if they say it can downgrade those
+  // collaboration tags.
+  UserX recognizesAsAuthority("DevicePolicyAuthority") :-
+    isCollaborationTag(tagX),
+    UserX says "DevicePolicyAuthority" canSay downgrades(tagX).
```
[the remark this answers is not in the document]
Comment (markww) [reply src/analysis/souffle/examples/multimic/multimic.authlogic:91]: Well, that's an interesting question. The microphone is sitting in this room, listening to what it thinks of as `userSpeech` all the time. As users come in and out, their personal information mingles with the `userSpeech` that this microphone is always gathering. As the microphone (and its associated peripherals) is the permanent part of this trust network; the users are ephemeral guests that dip in and out of it. From a "draw a graph of the network perspective", I was thinking it made more sense to think of the microphone as the owner of the `userSpeech` tag, but that it accommodated users that may have their privacy impacted by its data collection temporary shared ownership over that tag to help guide its behavior.

Imagine that there are no users in the room but an old phone message recorder that plays voice messages as it records them. The owner has made a little information-age Rube Goldberg machine to have its microphone send them transcripts of the messages that come in. Clearly anyone leaving a message on the machine has consented to being recorded. If the microphone does not own its concept of user speech, can it operate in this situation in which no policies are in the room?

I know that's a bit of a contrived thought experiment, but the issue is basically that the least privacy infringing listening that the microphone can do on the `userSpeech` aspect is record when no users are in the room. If the microphone does not own the `userSpeech` tag and is granted that by guest policies, it cannot record an empty room due to insufficient permissions, which seems incorrect.

File: src/analysis/souffle/examples/multimic/multimic.authlogic [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+  sensitiveDeviceTag("userSpeech").
+
+  ownsTag(tagX) :- sensitiveDeviceTag(tagX).
+
+  "UserC" hasVoiceRecordingPolicy("Never").
+
+  // This privacy-sensitive user only recognizes the authority of device
+  // authorities if they claim to notify when overriding your policy
+  // preferences.
+  PolciyAuthorityX canSay downgrades(tagX) :-
```
[the remark this answers is not in the document]
Comment (markww) [reply src/analysis/souffle/examples/multimic/multimic.authlogic:59]: Adding another user to this scenario may or may not make sense for this.

The scenario as originally described had some user that did not want their voice recorded entering within range of a smartmic that was performing text to speech. In the scenario, the mic could do two separate things:

1. Stop recording until they leave
2. Continue recording and send them a notification that their preferences could not be met

Adding another guest user policy, therefore, would not really illustrate the important part of the scenario, in which a guest user would have a preference that conflicted with the operation of the service and the service would decide how to handle the situation. If you were thinking User D as an alternative owning user with a different override-versus-stop preference, however, that could be good for expanding this example.

File: src/analysis/souffle/examples/multimic/multimic.authlogic [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+// UserB is a guest user. UserB is somewhat privacy sensitive. They want their
+// voice recorded only as a transcript and will only recognize authorities from
+// companies that they trust.
+"UserB" says {
+  // UserB has a set of sensitive device tags, which it always wants a say in.
+  sensitiveDeviceTag("userSpeech").
+
+  // UserB has a set of trusted companies that they have ongoing relationships
+  // with.
+  trustedCompanies("AWellKnownTechCompany").
```
[the remark this answers is not in the document]
Comment (markww) [reply src/analysis/souffle/examples/multimic/multimic.authlogic:19]: Sounds good. I was trying to make it clear that the mechanism could be generalized beyond one company stated explicitly, but perhaps that is obvious. I'll just get rid of the `trustedCompany` bit.

File: src/analysis/souffle/examples/multimic/multimic.arcs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  // The delegator claims that it downgrades the permissions to use storage
+  // from the full permissions to some less powerful delegated permissions.
+  claim outPermissions is not fullStoragePermissions and is delegatedPermissions
+  // Note: this is to halt if we are not allowed to downgrade permissions.
+  // We actually aren't allowed to do this in Arcs as it only allows checks on
+  // inputs! We should allow this in our new policy language.
+  // check outPermissions is not fullStoragePermissions
+
+// Represents the CPU within the mic.
+particle MicCompute
```
Comment (bgogul) [inline src/analysis/souffle/examples/multimic/multimic.arcs:22]: Call this `MicASR`?

File: src/analysis/souffle/examples/multimic/multimic.authlogic [hunk 5]
```diff
@@ -0,0 +1,132 @@
+// UserA is the device owner. UserA has a high level of trust in their own
+// device.
+"UserA" says {
+  "UserA" hasVoiceRecordingPreference("AlwaysOK").
```
Comment (bgogul) [inline src/analysis/souffle/examples/multimic/multimic.authlogic:4]: Is this valid syntax? @aferr

File: src/analysis/souffle/examples/multimic/multimic.authlogic [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+  "DevicePolicyAuthority" canSay downgrades(tagX) :- isTag(tagX).
+
+   "UserA" criticalDeviceServiceOverride("TextToSpeech").
```
Comment (bgogul) [inline src/analysis/souffle/examples/multimic/multimic.authlogic:7]: What does this mean?

File: src/analysis/souffle/examples/multimic/multimic.authlogic (hunk 5 above)
Comment (aferr) [reply src/analysis/souffle/examples/multimic/multimic.authlogic:4]: Yes: see [this paragraph](https://github.com/google-research/raksha/tree/main/rust/tools/authorization-logic#attributes-and-canactas) (especially the first sentence, but the rest will make it clear why this is different from just `hasVoiceRecordingPreference("UserA", "AlwaysOK")`)

File: src/analysis/souffle/examples/multimic/multimic.authlogic [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+  "UserA" canActAs "DeviceOwner".
+
+  // A collaboration tag can be influenced by PolicyCollaborators. The policy
+  // authority has one tag that it wishes to allow collaboration on,
+  // userSpeech. You could imagine this being a list of facts specifying
+  // many collaboration tags.
+  isCollaborationTag("userSpeech").
+
+  // The policy authority claims to own collaboration tags.
+  ownsTag(tagX) :- isCollaborationTag(tagX).
```
[the remark this answers is not in the document]
Comment (aferr) [reply src/analysis/souffle/examples/multimic/multimic.authlogic:80]: I think how to handle the fact that the users of the system are not known a priori and whether or not ownership of tags is moved to the other are orthogonal. I think to handle use-cases like this one involving multiple devices that enter/exit the system at runtime, we will need a way to send, inspect, and respond to authorization logic facts at runtime. When a new user joins, the user's device will probably need to send a set of facts to the microphone, and the microphone's policy code will depend on these facts. I think whether this set of facts passes ownership of the tag to the microphone is a separate concern. (I now understand what you are trying to express, though and can help flesh this out if that's useful).

**Comment by markww** [comment]
This is long out of date, having been superseded by later PRs. Closing.

**Pull request closed without merging**